src/beamlime/constructors/providers.py

```diff
@@ -0,0 +1,184 @@
+            raise err
+
+
+class ArgSubProviders:
+    def __init__(self, _constructor: Callable) -> None:
+        from .inspectors import collect_argument_dep_specs
+
+        self._subproviders = {
+            arg_name: _prov
+            for arg_name, arg_spec in collect_argument_dep_specs(_constructor).items()
+            if (_prov := _find_arg_provider(arg_spec)) != UnknownProvider
+        }
+
+    def build_arguments(self) -> Dict[str, Any]:
+        return {arg_name: _prov() for arg_name, _prov in self._subproviders.items()}
+
+
+class Provider:
+    """
+    Function wrapper that provides certain type of product.
+    """
+
+    def __new__(
+        cls, _provider_call: Union[Constructor, Provider], /, *args, **kwargs
+    ) -> None:
+        if isinstance(_provider_call, Provider):
+            # Nested ``Provider`` is avoided similar to ``partial``.
+            return Provider(_provider_call.constructor, *args, **kwargs)
+        elif isinstance(_provider_call, partial) and isinstance(
+            _provider_call.func, Provider
+        ):
+            partial_func = partial(
+                _provider_call.func.constructor,
+                *_provider_call.args,
+                **_provider_call.keywords,
+            )
+            return Provider(partial_func, *args, **kwargs)
+        else:
+            return super().__new__(Provider)
+
+    def __init__(
+        self, _provider_call: Union[Constructor, Provider], /, *args, **kwargs
+    ) -> None:
+        from .inspectors import get_product_spec
+
+        if isinstance(_provider_call, Provider):
+            ...
+        elif isinstance(_provider_call, partial) and isinstance(
+            _provider_call.func, Provider
+        ):
+            ...
+        else:
+            if args or kwargs:
+                _constructor = partial(_provider_call, *args, **kwargs)
+            else:
+                _constructor = _provider_call
+
+            self._constructor: Constructor = _constructor
+            self.product_spec = get_product_spec(_constructor)
+
+    @property
+    def call_name(self) -> str:
+        return ".".join((self.constructor.__module__, self.constructor.__qualname__))
+
+    @property
+    def constructor(self) -> Constructor:
+        return self._constructor
+
+    def can_provide(self, product_type: Union[ProductType, ProductSpec]):
+        """Check if the given ``product_type`` can be supported by this provider."""
+        from .inspectors import ischildproduct
+
+        return ischildproduct(self.product_spec, ProductSpec(product_type))
+
+    def __call__(self, *args, **kwargs):
+        """Build and return the product with attribute dependencies injected."""
+        # TODO: Split this into two steps, build chain of sub-providers
+        # and call the providers.
+        _constructor: Constructor
+
+        if args or kwargs:
+            _constructor = partial(self.constructor, *args, **kwargs)
+        else:
```

Comment:
The reason why I tried to inject the arguments/attributes here was because I wanted to use ``Provider`` as a function wrapper, so that it can just be called with the arguments as well, but...

Now I think it's not a very useful interface since it has the function call as a member variable anyways.

So this will be moved to ``Binder`` __getitem__ instead and the Provider class will have much less responsibility.

Reply:
This one is updated in the next commits.